scipy/stats/_stats_py.py

```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+
+        # Naming: variables ending in bn/b are for "between treatments", wn/w are
+        # for "within treatments"
+        ssbn = ssbn - normalized_ss
+        sswn = sstot - ssbn
+        dfbn = num_groups - 1
+        dfwn = bign - num_groups
+        msb = ssbn / dfbn
+        msw = sswn / dfwn
+        with np.errstate(divide='ignore', invalid='ignore'):
+            f = msb / msw
 
-    sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+        prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
 
-    ssbn = 0
-    for sample in samples:
-        smo_ss = _square_of_sums(sample - offset, axis=axis)
-        ssbn = ssbn + smo_ss / sample.shape[axis]
-
-    # Naming: variables ending in bn/b are for "between treatments", wn/w are
-    # for "within treatments"
-    ssbn = ssbn - normalized_ss
-    sswn = sstot - ssbn
-    dfbn = num_groups - 1
-    dfwn = bign - num_groups
-    msb = ssbn / dfbn
-    msw = sswn / dfwn
-    with np.errstate(divide='ignore', invalid='ignore'):
-        f = msb / msw
+    # if population variances are not equeal, perform Welch ANOVA
+    # the related issue: https://github.com/scipy/scipy/issues/11122
+    else:
+        # calculate basic statistics for each sample
+        samples_arrs = np.array(samples)
+
+        # adjust the shape of the data based on the arrary dimension for the calculation
+        # along with input axis
+        if samples_arrs.ndim <= 2:
+            samples_arrs = samples_arrs.T
+        else:
+            samples_arrs = samples_arrs.transpose(
+                *range(1, samples_arrs.ndim - 1), 0, -1
+            )
+
+        means = samples_arrs.mean(axis=axis)
+        ns = np.apply_along_axis(np.size, axis=axis, arr=samples_arrs)
+        k = np.apply_along_axis(np.shape, axis=axis, arr=ns)
+
+        vars_ = np.var(samples_arrs, axis=axis, ddof=1)
+        # calculate weight by number of data and varianc
+        ws = ns / vars_
+
+        # calculate adjusted grand mean
+        mean_prime = (np.sum(ws * means, axis=axis, keepdims=True) /
+                      np.sum(ws, axis=axis, keepdims=True))
+
+        numerator = (np.sum(ws * (means - mean_prime)**2, axis=axis, keepdims=True)
+                     / (k - 1))
+        denominator = (
+                1 + 2 * (k - 2) / (k ** 2 - 1) *
+                np.sum((1 / (ns - 1)) *
+                       (1 - ws / np.sum(ws)) **2,
+                       axis=axis, keepdims=True)
+        )
+        f = numerator / denominator
+
+        # adjusted degree of freedom
+        df = (
+                (k ** 2 - 1) /
+                (3 * np.sum((1 / (ns - 1)) *
+                            (1 - ws / np.sum(ws, axis=axis, keepdims=True)) **2,
+                            axis=axis, keepdims=True))
+        )
+
+        prob = stats.f.sf(f, k - 1, df)
 
-    prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
+        f = np.squeeze(f)
+        prob = np.squeeze(prob)
```

Comment:
This is typically not required; perhaps the last reducing operations should not have `keepdims=True`? If it *is* required, specify which is the singleton `axis` to be eliminate. 

Adding the test case to `test_axis_nan_policy.py` should catch bugs like this.